pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.ast
+
+import net.sourceforge.pmd.annotation.InternalApi
+import net.sourceforge.pmd.lang.apex.ApexParserOptions
+import net.sourceforge.pmd.lang.ast.ParseException
+import net.sourceforge.pmd.lang.ast.SourceCodePositioner
+
+import com.google.summit.ast.CompilationUnit
+import com.google.summit.ast.Identifier
+import com.google.summit.ast.Node
+import com.google.summit.ast.TypeRef
+import com.google.summit.ast.declaration.ClassDeclaration
+import com.google.summit.ast.declaration.EnumDeclaration
+import com.google.summit.ast.declaration.InterfaceDeclaration
+import com.google.summit.ast.declaration.TriggerDeclaration
+import com.google.summit.ast.declaration.TypeDeclaration
+import com.google.summit.ast.modifier.KeywordModifier
+import com.google.summit.ast.modifier.Modifier
+
+@Deprecated("internal")
+@InternalApi
+@Suppress("DEPRECATION")
+class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    private val sourceCodePositioner = SourceCodePositioner(sourceCode)
+
+    /** Builds and returns an [ApexNode] AST corresponding to the given [root] node. */
+    fun buildTree(root: CompilationUnit): ApexRootNode<TypeDeclaration> {
+        // Build tree
+        val result =
+            build(root, parent = null) as? ApexRootNode<TypeDeclaration>
+                ?: throw ParseException("Unable to build tree")
+
+        // Call additional methods
+        callAdditional(result)
+
+        return result
+    }
+
+    /** Calls additional methods for each node in [root] using a post-order traversal. */
+    private fun callAdditional(root: ApexNode<*>) =
```

Comment:
Any thoughts on a more specific name for this? Maybe... post order visitor? finalize nodes?